GPU/High/GLES/HighGLES.cpp

```diff
@@ -0,0 +1,435 @@
+	//
+	// Pass 1: Decode all the textures. This is done first so that the GL driver can
+	// upload them in the background (if it's that advanced) while we are decoding vertex
+	// data and preparing the draw calls.
+
+	int start = 0;
+	int end = packet->numCommands;
+	// Local mirror arrays. Should possible put it all in a struct instead and attach it
+	// using the userdata pointer.
+	TexCacheEntry *tex[ARRAY_SIZE(packet->texture)];
+
+	for (int i = 0; i < packet->numTexture; i++) {
+		const TextureState *texState = packet->texture[i];
+		const ClutState *clut = texState->clutStateIndex == INVALID_STATE ? nullptr : packet->clut[texState->clutStateIndex];
+		tex[i] = textureCache_->GetTexture(texState, clut);
+	}
+
+	short primCount[8] = {0};
+	short idxCount[4] = {0};
+	for (int i = start; i < end; i++) {
+		const Command *cmd = &packet->commands[i];
+		if (cmd->type != CMD_DRAWPRIM) {
+			continue;
+		}
+		primCount[cmd->draw.prim]++;
+		idxCount[(cmd->draw.vtxformat & GE_VTYPE_IDX_MASK) >> GE_VTYPE_IDX_SHIFT]++;
+		if (!(cmd->draw.enabled & ENABLE_TEXTURE))
+			continue;
+	}
+
+	// bool softwareTransform = primCount[GE_PRIM_RECTANGLES] > 0;
+
+	// Pass 2: Allocate a buffer and decode all the vertex data into it.
+	for (int i = start; i < end; i++) {
+		const Command *cmd = &packet->commands[i];
+		if (cmd->type != CMD_DRAWPRIM)
+			continue;
+
+		// VertexDecoder *dec = GetVertexDecoder(cmd->draw.vtxformat);
+		if (cmd->draw.vtxformat & GE_VTYPE_IDX_NONE) {
+			// Decode it all.
+			// dec->DecodeVerts(
+		}
+	}
+
+	// Pass 3: Fetch shaders, perform the draws.
+	int curFramebuf = -1;
+
+	for (int i = start; i < end; i++) {
+		const Command *cmd = &packet->commands[i];
+		if (curFramebuf != cmd->draw.framebuf) {
+			ApplyFramebuffer(packet, cmd);
+			curFramebuf = cmd->draw.framebuf;
+		}
+	}
+}
+
+void HighGpu_GLES::SetDisplayFramebuffer(u32 framebuf, u32 stride, GEBufferFormat format) {
+	framebufferManager_->SetDisplayFramebuffer(framebuf, stride, format);
+}
+
+VertexDecoder *HighGpu_GLES::GetVertexDecoder(u32 vtype) {
+	auto iter = decoderMap_.find(vtype);
+	if (iter != decoderMap_.end())
+		return iter->second;
+	VertexDecoder *dec = new VertexDecoder();
+	dec->SetVertexType(vtype, decOptions_, decJitCache_);
+	decoderMap_[vtype] = dec;
+	return dec;
+}
+
+// So much data to fetch to make the heuristic happy...
+void HighGpu_GLES::ApplyFramebuffer(const CommandPacket *packet, const Command *cmd) {
+	FramebufferHeuristicParams fb;
+	const FramebufState *fbState = packet->framebuf[cmd->draw.framebuf];
+	const RasterState *raster = packet->raster[cmd->draw.raster];
+	fb.fb_address = fbState->colorPtr;
+	fb.fb_stride = fbState->colorStride;
+	fb.fb_addr = fb.fb_address;
+	fb.z_address = fbState->depthPtr;
+	fb.z_stride = fbState->depthStride;
+	fb.fmt = (GEBufferFormat)(fbState->colorFormat);
+	// TODO: Maybe the viewport is such an important hint for this that we should always load it,
+	// even when drawing in throughmode?
```

Comment:
It really is important.  You're more than welcome to make the heuristics better, and they are ugly, but they do catch the common cases we can know about.  I don't like this `hasViewportAndRegion` because it just means we will size things wrong, and may even end up resizing framebufers.

Ultimately, we could potentially just make all framebuffers unnecessarily large.  The downsides of this are:
1. It can use a lot more VRAM, especially at higher IR.
2. It seems to have a performance penalty, at least in some cases.
3. We have to make sure we never clut/colorconv/etc. the full buffer.
4. If we're not tracking the estimated height anymore, there will be (more?) cases where we incorrectly texture from a framebuffer instead of raw VRAM, unless we catch those all via replacements.  There will be a speed penalty either way.

-[Unknown]